components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java

```diff
@@ -83,13 +83,29 @@ public ServiceURL build() throws URLBuilderException {
         String protocol = fetchProtocol();
         String proxyHostName = fetchProxyHostName();
         String internalHostName = fetchInternalHostName();
+        String authenticationEndpointHostName = fetchAuthenticationEndpointHostName();
+        String authenticationEndpointPath = fetchAuthenticationEndpointPath();
+        String recoveryEndpointHostName = fetchRecoveryEndpointHostName();
+        String recoveryEndpointPath = fetchRecoveryEndpointPath();
         int proxyPort = fetchPort();
         int transportPort = fetchTransportPort();
         String tenantDomain = StringUtils.isNotBlank(tenant) ? tenant : resolveTenantDomain();
         String proxyContextPath = ServerConfiguration.getInstance().getFirstProperty(PROXY_CONTEXT_PATH);
         String resolvedFragment = buildFragment(fragment, fragmentParams);
         String urlPath = getResolvedUrlPath(tenantDomain);
 
+        if (StringUtils.isNotBlank(urlPath)) {
+            if (authenticationEndpointHostName != null && authenticationEndpointPath != null &&
+                    urlPath.contains(authenticationEndpointPath)) {
+                return new ServiceURLImpl(protocol, authenticationEndpointHostName, internalHostName, proxyPort,
+                        transportPort, tenantDomain, proxyContextPath, urlPath, parameters, resolvedFragment);
+            }
+            if (recoveryEndpointHostName != null && recoveryEndpointPath != null &&
+                    urlPath.contains(recoveryEndpointPath)) {
+                return new ServiceURLImpl(protocol, recoveryEndpointHostName, internalHostName, proxyPort,
+                        transportPort, tenantDomain, proxyContextPath, urlPath, parameters, resolvedFragment);
+            }
+        }
```

Comment:
IMO this should be handled at the places where the recovery endpoint path and the authentication endpoint path are there in the URLs. Instead of handling it at the ServiceURLBuilder, can't we include this logic at those specific places? WDYT @mevan-karu @mefarazath ?

Reply:
@ChamathNS but according to the current implementation of the ServiceURLBuilder, we can't configure the hostname from outside. Hostname resolving is done internally by the ServiceURLBuilder.